### ISSUE TITLE
Default values in polymorphic function definitions

### ISSUE THREAD

**Issue by aptmcl** [racket/racket#1130]
The following example typeckecks

``` racket
(define #:forall (A) (foo [f : (Real -> A)] [b : Boolean]) : (Listof A)
  (list (f 0.1) (f 0.2)))
```

but the following example generates a lot of complaints:

``` racket
(define #:forall (A) (foo [f : (Real -> A)] [b : Boolean #t]) : (Listof A)
  (list (f 0.1) (f 0.2)))
```

Welcome to DrRacket, version 6.3.0.3--2015-11-06(58c919c/a) [3m].
Language: typed/racket; memory limit: 4096 MB.
. Type Checker: parse error in type;
 type name `A' is unbound in: A
. Type Checker: parse error in type;
 type name`A' is unbound in: A
. Type Checker: parse error in type;
 type name `A' is unbound in: A
. Type Checker: parse error in type;
 type name`A' is unbound in: A
. Type Checker: parse error in type;
 type name `A' is unbound in: A
. Type Checker: parse error in type;
 type name`A' is unbound in: A
. Type Checker: parse error in type;
 type name `A' is unbound in: A
. Type Checker: parse error in type;
 type name`A' is unbound in: A
. Type Checker: Summary: 8 errors encountered in:
  A
  A
  A
  A
  A
  A
  A
  A

**Comment by harrisi** [comment]
Disclaimer: I don't know much of anything about anything.

I'm interested in this because I can get this to type check with a default argument, but I don't understand how I would use that or how it could be well-typed. I can't possibly call that and use the default value (unless I call it with the default value anyway) and have it still be well-typed.

```racket
(: foo (All (A) (-> (Real -> A) Boolean (Listof A))))
(define (foo f [b #t])
  (list (f 0.1) (f 0.2)))
```

This type checks just fine, but I still have to call it with two arguments:

```
> (foo add1 #t)
- : (Listof Real)
'(1.1 1.2)
> (foo add1)
. Type Checker: Polymorphic function `foo' could not be applied to arguments:
Wrong number of arguments - Expected 2, but got 1

Argument 1:
  Expected: (-> Real A)
  Given:    (-> Number Number)
Argument 2:
  Expected: Boolean
  Given:    -none-
 in: (foo add1)
```

I very well may be missing something, but without having an additional type or struct involved (see: [Haskell](http://stackoverflow.com/questions/7781096/is-there-a-better-way-to-have-optional-arguments-in-haskell), [Racket](http://stackoverflow.com/questions/10871275/how-do-you-initialize-a-struct-with-default-values-in-typed-racket), [more Haskell](http://neilmitchell.blogspot.com/2008/04/optional-parameters-in-haskell.html)), I don't see how I can have the function application adhere to the type you defined.

So in conclusion, I think the error message given is a bit odd, but also I think that the expected behavior isn't well-defined in this issue.

**Comment by samth** [comment]
@harrisi The way to write your code is:

```
(: foo (All (A) (->* [(Real -> A)] [Boolean] (Listof A))))
(define (foo f [b #t])
  (list (f 0.1) (f 0.2)))
```

That's the best way to work around this issue, but there's a real problem with the scoping in `#:forall`.

**Comment by samth** [comment]
Fixed by racket/typed-racket#878.
